Objects/frameobject.c

```diff
@@ -53,22 +53,26 @@ static int
 framelocalsproxy_getkeyindex(PyFrameObject *frame, PyObject* key, bool read)
 {
     /*
-     * Returns the fast locals index of the key
+     * Returns -2 (!) if an error occurred; exception will be set.
+     * Returns the fast locals index of the key on success:
      *   - if read == true, returns the index if the value is not NULL
      *   - if read == false, returns the index if the value is not hidden
+     * Otherwise returns -1.
      */
 
-    assert(PyUnicode_CheckExact(key));
-
     PyCodeObject *co = _PyFrame_GetCode(frame->f_frame);
-    int found_key = false;
+
+    // Ensure that the key is hashable.
+    Py_hash_t key_hash = PyObject_Hash(key);
+    if (key_hash == -1) {
+        return -2;
+    }
 
     // We do 2 loops here because it's highly possible the key is interned
     // and we can do a pointer comparison.
     for (int i = 0; i < co->co_nlocalsplus; i++) {
         PyObject *name = PyTuple_GET_ITEM(co->co_localsplusnames, i);
         if (name == key) {
-            found_key = true;
```

Comment:
If you removed `found_key` check here, maybe just return `-1` if the condition does not meet. Otherwise it will go through the loop again and check something we already knew.

Reply:
The rest of the PR LGTM.

Reply:
Right, thanks for the catch!

Reply:
Could you check the change?
I'm still not too familiar with the internals of this mapping; I just found out (hopefully correctly) what the undocumented `CO_FAST_HIDDEN` does.

Reply:
Capturing my understanding of the `CO_FAST_HIDDEN` handling, since it's genuinely not clear:

* in Python 3.12, the iteration variables for inlined comprehensions are visible for reads while the comprehension(s) is(/are) running, but not writable, and not visible outside the comprehension(s)
* in Python 3.13, all of those properties remain true when fast locals proxy instances are emitted during iteration (writes to hidden variables go into the frame's "extra locals" storage rather than modifying the comprehension iteration variable)

Future version could revisit that design if debugger developers ask for it to change, but live editing loop variables seems sufficiently fraught with peril that I'll be genuinely surprised if anyone ever notices this small discrepancy between module level comprehensions and function level ones.